shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/TemporaryTerminalRewriter.kt

```diff
@@ -0,0 +1,254 @@
+
+        return if (plausibleTruncatedPatterns.size < 2) {
+            plausibleTruncatedPatterns.singleOrNull()
+        } else {
+            // if there are multiple truncated patterns with the right direction and stops,
+            // hopefully there's only one that's actually on the schedule
+            plausibleTruncatedPatterns.singleOrNull { truncatedPattern ->
+                schedulesByRoute[fullPattern.routeId].orEmpty().any {
+                    it.routePattern() == truncatedPattern
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks each typical pattern and child stop at [stopPatterns] for truncation, and merges full
+     * patterns into their truncated [NearbyStaticData.StaticPatterns].
+     */
+    fun truncatePatternsAtStop(
+        stopPatterns: NearbyStaticData.StopPatterns
+    ): NearbyStaticData.StopPatterns {
+        for (fullPattern in stopPatterns.patterns.flatMap { it.patterns }) {
+            if (fullPattern.typicality != RoutePattern.Typicality.Typical) continue
+            val fullPatternStopIds =
+                globalData.trips[fullPattern.representativeTripId]?.stopIds ?: continue
+
+            for (stopId in stopPatterns.allStopIds) {
+                if (!fullPatternStopIds.contains(stopId)) continue
+
+                truncatedPatternByFullPatternAndStop[Pair(fullPattern.id, stopId)] =
+                    truncatedPattern(fullPattern, fullPatternStopIds, stopId)?.id
+            }
+        }
+
+        val collapsedPatterns =
+            stopPatterns.patterns
+                .groupBy { staticPatterns ->
+                    staticPatterns.patterns.map { pattern ->
+                        val truncatedPatternId =
+                            stopPatterns.allStopIds
+                                .mapNotNull { stopId ->
+                                    truncatedPatternByFullPatternAndStop[Pair(pattern.id, stopId)]
+                                }
+                                .singleOrNull()
+                        truncatedPatternId ?: pattern.id
+                    }
+                }
+                .values
+                .map { patternsList ->
+                    if (patternsList.size == 1) return@map patternsList.single()
+                    val patternsCorrectFirst =
+                        patternsList.sortedBy {
+                            val isTruncated =
+                                it.patterns.any { pattern ->
+                                    truncatedPatternByFullPatternAndStop.containsValue(pattern.id)
+                                }
+                            if (isTruncated) 0 else 1
+                        }
+                    patternsCorrectFirst
+                        .reduce { acc, nextPatterns ->
+                            acc.copy(patterns = acc.patterns + nextPatterns.patterns)
+                        }
+                        .let { it.copy(patterns = it.patterns.sorted()) }
+                }
```

Comment:
I'm confused by what this is doing - could it be broken down differently or commented?

Reply:
Broke this down slightly differently and gave the `groupBy` lambda a name.